File: c/linked-list/linked_list.c
#include "linked_list.h"

#include <stdlib.h>

struct list_node {
    struct list_node *prev, *next;
    ll_data_t data;
};

struct list {
    struct list_node *first, *last;
};

static struct list_node* new_node(ll_data_t data) {
    struct list_node* n = malloc(sizeof(struct list_node));
    n->prev = NULL;
    n->next = NULL;
    n->data = data;
    return n;
}

struct list* list_create(void) {
    struct list* l = malloc(sizeof(struct list));
    l->first = NULL;
    l->last = NULL;
    return l;
}

size_t list_count(const struct list* l) {
    size_t count = 0;
    struct list_node* node = l->first;
    while (node != NULL) {
        count++;
        node = node->next;
    }
    return count;
}

void list_push(struct list* l, ll_data_t item_data) {
    struct list_node* n = new_node(item_data);

    if (l->first == NULL || l->last == NULL) {
        l->first = n;
        l->last = n;
    } else {
        l->last->next = n;
        n->prev = l->last;
        l->last = n;
    }
}

ll_data_t list_pop(struct list* l) {
    struct list_node* n = l->last;
    if (n == NULL) return 0;
    ll_data_t data = n->data;
    l->last = l->last->prev;
    if (l->last != NULL) {
        l->last->next = NULL;
    } else {
        l->first = NULL;
    }
    free(n);
    return data;
}

void list_unshift(struct list* l, ll_data_t item_data) {
    struct list_node* n = new_node(item_data);

    if (l->first == NULL || l->last == NULL) {
        l->first = n;
        l->last = n;
    } else {
        l->first->prev = n;
        n->next = l->first;
        l->first = n;
    }
}

ll_data_t list_shift(struct list* l) {
    struct list_node* n = l->first;
    if (n == NULL) return 0;
    ll_data_t data = n->data;
    l->first = l->first->next;
    if (l->first != NULL) {
        l->first->prev = NULL;
    } else {
        l->last = NULL;
    }
    free(n);
    return data;
}

void list_delete(struct list* l, ll_data_t data) {
    struct list_node* n = l->first;
    while (n != NULL && n->data != data) {
        n = n->next;
    }

    if (n == NULL) return;
    if (n->prev != NULL) n->prev->next = n->next;
    if (n->next != NULL) n->next->prev = n->prev;
    if (l->first == n) l->first = n->next;
    if (l->last == n) l->last = n->prev;
    free(n);
}

void list_destroy(struct list* l) {
    struct list_node* n = l->first;
    while (n != NULL) {
        struct list_node* nxt = n->next;
        free(n);
        n = nxt;
    }
    free(l);
}

File: zig/linked-list/linked_list.zig
pub fn LinkedList(comptime T: type) type {
    return struct {
        pub const Node = struct {
            prev: ?*Node = null,
            next: ?*Node = null,
            data: T,
        };

        first: ?*Node = null,
        last: ?*Node = null,
        len: usize = 0,

        pub fn push(self: *LinkedList(T), node: *Node) void {
            node.prev = self.last;
            if (self.last) |last_node|
                last_node.next = node
            else
                self.first = node;
            self.last = node;
            self.len += 1;
        }

        pub fn pop(self: *LinkedList(T)) ?*Node {
            if (self.last) |last_node| {
                self.last = last_node.prev;
                self.len -= 1;
                if (self.len == 0) self.first = null;
                return last_node;
            }
            return null;
        }

        pub fn shift(self: *LinkedList(T)) ?*Node {
            if (self.first) |first_node| {
                self.first = first_node.next;
                self.len -= 1;
                if (self.len == 0) self.last = null;
                return first_node;
            }
            return null;
        }

        pub fn unshift(self: *LinkedList(T), node: *Node) void {
            node.next = self.first;
            if (self.first) |first_node|
                first_node.prev = node
            else
                self.last = node;
            self.first = node;
            self.len += 1;
        }

        pub fn delete(self: *LinkedList(T), node: *Node) void {
            var curr: ?*Node = self.first;
            while (curr != null) {
                const cur: *Node = curr.?;
                if (cur.data == node.data) {
                    const prev = cur.prev;
                    const next = cur.next;
                    if (self.first.?.data == node.data) self.first = next;
                    if (self.last.?.data == node.data) self.last = prev;
                    if (prev != null) prev.?.next = next;
                    if (next != null) next.?.prev = prev;
                    self.len -= 1;
                    return;
                }
                curr = cur.next;
            }
        }
    };
}
